### ISSUE TITLE
Allow choosing a style setting for Invoke-DbatoolsFormatter

### ISSUE THREAD

**Issue by jayvdb** [dataplat/dbatools#5829]
### Summary of new feature

`Invoke-DbatoolsFormatter` currently only uses `CodeFormattingOTBS` setting, which is hard-coded inside the code.

These settings have some odd inconsistencies in the definition itself, which are only exposed when the setting definition is used with the linter instead of the reformatter.  See https://github.com/PowerShell/PSScriptAnalyzer/issues/769

As a result, I can not set my project to use `CodeFormattingOTBS` as-is, and I cant create a derivative that inherits and slightly tweaks the existing definition of `CodeFormattingOTBS`.  c.f. https://gitter.im/PowerShell/PSScriptAnalyzer?at=5d12d0523e73272c54c191e4

I need a new definition file, which will be a saner version of `CodeFormattingOTBS`, and I really want to be able to use that setting file consistently everywhere, which means pass that setting into `Invoke-DbatoolsFormatter`.

https://gitlab.com/coala/mobans/merge_requests/129/diffs#6dd0f3b41db66c66ef3e612aa43b49453d555045

### Latest version of dbatools as of writing

dbatools 1.0.5

**Comment by potatoqualitee** [comment]
Thanks for the report, Jay. @niphlod can you take a look?

**Comment by niphlod** [comment]
ahum.... lemme be straight. What Invoke-DbatoolsFormatter has been created for is to be able to format consistently all our codebase.
There have been countless hours passed discussing how to place a bracket or how to properly case begin, process and end block and, really, it didn't really close the deal (not even creating a wiki about codestyle).
Requiring PR reviewers to remember and adhere to all the rules was a major PITA, so we settled for a good standard and, given what we use is the same that VSCode use, it was a win-win.
Long story short, we favor consistency over nitty-gritty details.
The moment we don't like OTSB, we discuss on another "standard", we code it into VSCode profiles and into Invoke-DbaToolsFormatter, and we invoke it, one time, then commit all.
From there on, all the code will need to adhere to the new formatting standard but, I repeat, it is consistent and ENFORCED to be, because we have also CI checking for it at each and every commit.

That being said, if you want to discuss a new style for dbatools, join us on slack and propose a new one.
I promise everyone will chime in ;-)
Invoke-DbaToolsFormatter won't EVER take a style as a parameter, because it was made to force ONE and only one style, and just for dbatools (really, it's just a glorified wrapper over PSSA's Invoke-Formatter).
We DO know there's an issue with PSSA > 1.17.1 with OTSB but we're just waiting for it to be fixed (now that you remember me, it's taking unusually long).

**Comment by potatoqualitee** [comment]
i got the feeling that Jay is using your 💣  tool to fix his own non-dbatools related code.

**Comment by wsmelton** [comment]
Yeah, to be clear the purpose of this command is for our module and our module alone. We didn't create this command to help other modules or code formatting needs; nor do we plan on changing it. It is meant to help the code reviewers and maintainers of this module.

**Comment by potatoqualitee** [comment]
But if he is using it to fix his own code, that is an unexpected but cool use-case ❄️ . Throwing in one param for style does not seem to be a big burden for us.

**Comment by jayvdb** [comment]
Ya, I am mis-using your tool, because the raw formatter tool is unusable without a wrapper.

I appreciate that you are using this to ensure your own codebase is nicely styled and avoiding all of those silly style discussions.  I am from @coala , and that is exactly the use case that we build tools for: enforcing style in CI.  However @coala has no support for Powershell atm, and I needed a linter/formatter for a very large chunk of PS1 needed to get our Windows support beefed up, so that we can start tackling .net based languages like PS1.  Chicken and egg problem for me.

I found your wrapper.  I looked high and low, and it is one of the best wrappers.  It helped me get started and I now have my own wrapper which handles the same things, in a slightly different way to be more efficient for my needs, and with fixes to the bugs/features I have been raising here.  I hope that the issues I have raised here have been generic improvements to your tool which also benefit your own community.

I raised this mostly because of this function is using OTSB, and if I am not mistaken it is defined upstream, and it is illogical.  See https://github.com/PowerShell/PSScriptAnalyzer#769  That is a two year old problem.  You are not noticing this problem because your pester rule only excludes a few rules explicitly (and you have an include list at [bin/PSScriptAnalyzerRules.psd1](https://github.com/sqlcollaborative/dbatools/blob/development/bin/PSScriptAnalyzerRules.psd1) and I am not sure how PSSA merges those, but I do know PSScriptAnalyzer also disables all of the coding style rules because they need to be configured.  If you tried to add the OTSB configuration to your pester invocation, you'll find it is impossible to satisfy its rules because of that issue.

If you read the discussion at https://gitter.im/PowerShell/PSScriptAnalyzer?at=5d12d0523e73272c54c191e4 , you'll see that to enforce your code style properly, you need to invoke PSScriptAnalyzer twice - one with the default rules, and once with a slightly modified CodeFormattingOTBS .  Or build a single unified settings file like I have done at https://gitlab.com/coala/mobans/merge_requests/129 

PSScriptAnalyzer finds some quite significant problems that Invoke-Formatter ignores because it doesnt yet know how to handle.  So my PSLint.ps1 uses both together for automated fixes + the errors where fixes are not available.

dbatools can do the same, but it needs to choose between one of the two illogical settings described at https://github.com/PowerShell/PSScriptAnalyzer#769

**Comment by niphlod** [comment]
aha, now I get it, thanks for the clarification. we kept things separate (formatting and syntax-checking) in order to be agile and not scare the newbies away.
You're right though (if I grasped the thread): if you ask scriptanalyzer to check for formatting (also), it clashes with the very format invoke-formatter creates for the same settings.
Need to put my head into this and possibly also the bug on formatter on > 1.17, maybe we'll find a way to "commonize" (or, even, just steal your findings).
We have to implement compatibilityrules anyway so a deep dive seems quite unavoidable.
At this point having a parameter that enables you to use your own settings file while dbatools provides its own (e.g. within the module directory) is not that big of a stretch

**Comment by andreasjordan** [comment]
Do we still want to work on this?

**Comment by wsmelton** [comment]
The use of the module itself isn't really required as all that users/contributors have to do now is use the format in VS Code+PS Extension. We have a PSSA configuration file in our project and have for years now.

**Comment by potatoqualitee** [comment]
this is a dbatools-only command intended to help our devs comply with our code base requirements. i'd suggest using AI to build a VS Code extension or cannibalizing this command (I built something recently for personal use, had a blast and it's useful)
